### PR TITLE
Enable install hook for k8s charms.

### PR DESCRIPTION
**Pull request by hmlanigan**
Part of uniter++ work.

### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
There is a trello card for the integration test.
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Enable install hook for k8s charms while skipping the install operation done by the uniter.  It is not required due to the k8s pods.  

Do not block storage hooks based on the install hook not having been run for CAAS units, other wise it will hang if there is storage.

## QA steps

Build juju for CAAS and bootstrap
```sh
make microk8s-operator-update
juju bootstrap microk8s mk8s
juju add-model test
```
Deploy this specially ubuntu charm to watch the install, start and update-status hooks being called.  Check the debug log for Hello with each.
```sh
docker pull ubuntu:latest
juju deploy $GOPATH/src/github.com/juju/juju/testcharms/charm-repo/kubernetes/ubuntu --resource ubuntu_image=ubuntu
```
Deploy a k8s charm with storage.
```sh
juju create-storage-pool operator-storage
juju create-storage-pool mariadb-pv
juju deploy cs:~juju/mariadb-k8s --storage database=mariadb-pv,10M
```
Once they're both up and running:
```sh
juju show-status-log mariadb-k8s/0
juju show-status-log ubuntu-k8s/0
juju debug-log --replay | grep ubuntu | grep juju-log
juju storage
```
Look for install -> leadership -> config-changed -> start -> database-storage-attached hooks to be called in order in the show-status-log output.  The storage-attached hook will only be on the mariadb-k8s unit.
## Documentation changes

Yes, a discourse post will be done related to the order of hooks For CAAS units based on this change.

## Bugs

https://bugs.launchpad.net/juju/+bug/1854635

### CODE REVIEW

File: worker/uniter/storage/resolver.go [hunk 1]
```diff
@@ -70,14 +65,24 @@ func (s *storageResolver) NextOp(
 		return nil, errors.Trace(err)
 	}
 
+	// Only IAAS models need to first run the install hook.
+	// For CAAS models, storage is specified in the pod config
+	// and mounted as the pod in started.
+	blockedWaitingForInstall := !localState.Installed && s.modelType == model.IAAS
```
Comment (hmlanigan) [inline worker/uniter/storage/resolver.go]: Changing this to !localState.Installed, didn't work, CAAS units with storage hand.

File: worker/uniter/storage/resolver.go [hunk 2]
```diff
@@ -70,14 +65,24 @@ func (s *storageResolver) NextOp(
...
 	switch {
 	case localState.Kind == operation.Continue:
 		// There's nothing in progress.
-		runStorageHooks = true
+		fallthrough
 	case blockedWaitingForInstall && localState.Kind == operation.RunHook && localState.Step == operation.Queued:
 		// The install operation completed, and there's an install
 		// hook queued. Run storage-attached hooks first.
+		fallthrough
+	case s.modelType == model.CAAS:
```
Comment (hmlanigan) [inline worker/uniter/storage/resolver.go]: Is this a bandaid and the fix is somewhere else?

**Comment by hmlanigan** [comment]
@wallyworld could you please take a look.  The change to the storage resolver makes me think a change is needed somewhere else instead, but darned if I know where.  TY!

**Comment by achilleasa** [comment]
@hmlanigan QA ok!

**Comment by wallyworld** [comment]
Not sure how QA passed :-)
Here's a diff with the required code changes. Just the code, still needs tests etc.
https://pastebin.ubuntu.com/p/bn5fvcmySc/

**Comment by hmlanigan** [comment]
@wallyworld, with the provided changed, the mariadb-k8s unit I deployed isn't running the start hook at all.  https://pastebin.canonical.com/p/z6STfrxgGt/

The code I used is pushed up now.

**Comment by wallyworld** [comment]
The change I did has a timing issue, so here's a tweak
https://pastebin.ubuntu.com/p/gxH9ddNzjS/

If the storage became ready between the initial config-changed but before started, the hook commit would wipe out the state put in place by the config-changed hook to chain the execution of the started hook.

**Comment by hmlanigan** [comment]
@wallyworld, we had the same idea on how to fix.  :-). Your explanation in the code is far better than mine.  ty.

**Comment by hmlanigan** [comment]
start hook is now called for CAAS units with storage.  including a unit test.

File: worker/uniter/storage/resolver.go [hunk 3]
```diff
@@ -70,21 +65,51 @@ func (s *storageResolver) NextOp(
...
 		runStorageHooks = true
-	case blockedWaitingForInstall && localState.Kind == operation.RunHook && localState.Step == operation.Queued:
-		// The install operation completed, and there's an install
-		// hook queued. Run storage-attached hooks first.
-		runStorageHooks = true
+	case localState.Kind == operation.RunHook && localState.Step == operation.Queued:
+		// There's a hook queued.
+		switch s.modelType {
+		case model.CAAS:
+			// For CAAS models, we wait until after the starte hook has run before
```
Comment (wallyworld) [inline worker/uniter/storage/resolver.go]: oops, typo

File: worker/uniter/storage/resolver.go [hunk 4]
```diff
@@ -70,21 +65,51 @@ func (s *storageResolver) NextOp(
...
+			// running storage-attached hooks since storage is provisioned only
+			// after the pod has been created.
+			//
+			// NB we must be careful here. The initial hook sequence is
+			// install->leader-elected->config-changed->started
+			// This chain is activated by each preceding hook queuing the next.
+			// If we allow a storage-attached hook to run before the start hook, we
+			// will potentially overwrite the state to initiate the next hook, start.
```
Comment (wallyworld) [inline worker/uniter/storage/resolver.go]: delete ", start"

File: worker/uniter/storage/resolver.go (hunk 3 above)
Comment (hmlanigan) [reply worker/uniter/storage/resolver.go]: done

File: worker/uniter/storage/resolver.go (hunk 4 above)
Comment (hmlanigan) [reply worker/uniter/storage/resolver.go]: done

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**